README.md
# chef-mailhog

[![CK Version](http://img.shields.io/cookbook/v/mailhog.svg)](https://supermarket.getchef.com/cookbooks/mailhog)

This cookbook installs [MailHog](https://github.com/mailhog/MailHog).

Attributes
-----

This cookbook contains the following attributes:

| Key                                              | Type    | Default                                                           | Description                                                                  |
|--------------------------------------------------|---------|-------------------------------------------------------------------|------------------------------------------------------------------------------|
| ['mailhog']['version']                           | String  | 0.2.0                                                             | Version of the MailHog to install                                            |
| ['mailhog']['install_method']                    | String  | binary                                                            | MailHog install method                                                       |
| ['mailhog']['binary']['url']                     | String  | nil                                                               | MailHog binary url                                                           |
| ['mailhog']['binary']['mode']                    | Integer | 0755                                                              | Access permissions to the MailHog binary file                                |
| ['mailhog']['binary']['path']                    | String  | /usr/local/bin/MailHog                                            | Location of the MailHog binary file                                          |
| ['mailhog']['binary']['prefix_url']              | String  | https://github.com/mailhog/MailHog/releases/download/v            | MailHog binary prefix url                                                    |
| ['mailhog']['binary']['checksum']['linux_386']   | String  | a72d1016b70964562c8a77a3b57637a77889ee61f3b22973e0a7beb17181d8da  | MailHog binary file checksum for linux_386                                   |
| ['mailhog']['binary']['checksum']['linux_amd64'] | String  | e8e9acb4fa4470f4d4c3a4bba312f335bfc28122ea723599531699f099b4c9a5  | MailHog binary file checksum for linux_amd64                                 |
| ['mailhog']['service']['owner']                  | String  | root                                                              | User that should own the mailhog service                                     |
| ['mailhog']['service']['group']                  | String  | root                                                              | Group that should own the mailhog service                                    |
| ['mailhog']['smtp']['ip']                        | String  | 0.0.0.0                                                           | Interface for SMTP server to bind to                                         |
| ['mailhog']['smtp']['port']                      | Integer | 1025                                                              | Port for SMTP server to bind to                                              |
| ['mailhog']['smtp']['outgoing']                  | String  | nil                                                               | JSON file defining outgoing SMTP servers                                     |
| ['mailhog']['api']['ip']                         | String  | 0.0.0.0                                                           | Interface for HTTP API server to bind to                                     |
| ['mailhog']['api']['port']                       | Integer | 8025                                                              | Port for HTTP API server to bind to                                          |
| ['mailhog']['ui']['ip']                          | String  | 0.0.0.0                                                           | Interface for HTTP UI server to bind to                                      |
| ['mailhog']['ui']['port']                        | Integer | 8025                                                              | Port for HTTP UI server to bind to                                           |
| ['mailhog']['ui']['web_path']                    | String  | ''                                                                | WebPath under which the ui is served (without leading or trailing slahes), e.g. 'mailhog'  |
| ['mailhog']['cors-origin']                       | String  | nil                                                               | If set, a Access-Control-Allow-Origin header is returned for API endpoints   |
| ['mailhog']['hostname']                          | String  | mailhog.example                                                   | Hostname to use for EHLO/HELO and message IDs                                |
| ['mailhog']['storage']                           | String  | memory                                                            | Set message storage: memory / mongodb / maildir                              |
| ['mailhog']['maildir_path']                      | String  | ''                                                                | Path for storage of messages in a file                                       |
| ['mailhog']['mongodb']['ip']                     | String  | 127.0.0.1                                                         | Host for MongoDB                                                             |
| ['mailhog']['mongodb']['port']                   | Integer | 27017                                                             | Port for MongoDB                                                             |
| ['mailhog']['mongodb']['db']                     | String  | mailhog                                                           | MongoDB database name for message storage                                    |
| ['mailhog']['mongodb']['collection']             | String  | messages                                                          | MongoDB collection name for message storage                                  |
| ['mailhog']['jim']['enable']                     | Boolean | false                                                             | Set to true to enable Jim                                                    |
| ['mailhog']['jim']['accept']                     | Float   | 0.99                                                              | Chance of accepting an incoming connection                                   |
| ['mailhog']['jim']['disconnect']                 | Float   | 0.005                                                             | Chance of randomly disconnecting a session                                   |
| ['mailhog']['jim']['linkspeed']['affect']        | Float   | 0.1                                                               | Chance of applying a rate limit                                              |
| ['mailhog']['jim']['linkspeed']['max']           | Integer | 10240                                                             | Maximum link speed (in bytes per second)                                     |
| ['mailhog']['jim']['linkspeed']['min']           | Integer | 1024                                                              | Minimum link speed (in bytes per second)                                     |
| ['mailhog']['jim']['reject']['auth']             | Float   | 0.05                                                              | Chance of rejecting an AUTH command                                          |
| ['mailhog']['jim']['reject']['recipient']        | Float   | 0.05                                                              | Chance of rejecting a RCPT TO command                                        |
| ['mailhog']['jim']['reject']['sender']           | Float   | 0.05                                                              | Chance of rejecting a MAIL FROM command                                      |

Usage
-----

Include the mailhog recipe to install MailHog on your system:
```chef
include_recipe "mailhog::default"
```

MailHog service is installed and managed via `runit`.

Requirements
------------

### Cookbooks:

* runit

### Platforms:

* Ubuntu
* Debian
* RHEL
* CentOS
* Fedora

License
-------

Copyright (c) 2015 Sergey Storchay, http://r8.com.ua
Modified 2016 Gleb Levitin, dkd Internet Service GmbH

Licensed under MIT:
http://raw.github.com/r8/chef-mailhog/master/LICENSE


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased
- Add support for maildir storage [[SeanSith](https://github.com/SeanSith)]

## [0.2.1] - (2020-09-23)
### Added
-  Add option for ui web path [[Alexj12](https://github.com/Alexj12)]

## [0.2.0] - (2016-05-31)
### Changed
- Update github url to use https [[daohoangson](https://github.com/daohoangson)]
- Set runit dependency to version 1.7.4  [[levitin](https://github.com/levitin)]

### Added
- Create spec tests [[levitin](https://github.com/levitin)]
- Extend mailhog configuration with command line options [[levitin](https://github.com/levitin)]
- Extend readme documentation with command line options  [[levitin](https://github.com/levitin)]

## [0.1.0] - (2015-03-23)

- Initial release


spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: mailhog
# Spec:: default
#
# Copyright (c) 2015 Sergey Storchay, All Rights Reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

require 'spec_helper'

describe 'mailhog::default' do

  context 'When all attributes are default, on an unspecified platform' do

    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      chef_run # This should not raise an error
    end

  end
end


test/integration/default/serverspec/default_spec.rb
require 'spec_helper'

describe 'mailhog::default' do

  describe file('/usr/local/bin/MailHog') do
    it { should exist }
    it { should be_file }
    it { should be_mode 755 }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
  end

  describe file('/etc/service/mailhog/run') do
    it { should exist }
    it { should be_file }
    it { should be_mode 755 }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
  end

  describe service('mailhog') do
    it { should be_running }
  end

  describe port(1025) do
    it { should be_listening.on('127.0.0.1').with('tcp') }
  end

  describe port(8025) do
    #see: https://github.com/mailhog/MailHog/issues/87
    it { should be_listening.on('::').with('tcp6') }
  end

end

recipes/binary.rb
#
# Cookbook Name:: mailhog
# Recipe:: default
#
# Copyright (c) 2015 Sergey Storchay, All Rights Reserved.
# Modified 2016 Gleb Levitin, dkd Internet Service GmbH
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

include_recipe 'runit::default'

arch = node['kernel']['machine'] =~ /x86_64/ ? 'amd64' : '386'

if node['mailhog']['binary']['url']
  binary_url = node['mailhog']['binary']['url']
  checksum = node['mailhog']['binary']['checksum']
else
  binary_url = "#{node['mailhog']['binary']['prefix_url']}#{node['mailhog']['version']}/MailHog_linux_#{arch}"
  checksum = node['mailhog']['binary']['checksum']["linux_#{arch}"]
end

# Download and install binary file
remote_file node['mailhog']['binary']['path'] do
  source binary_url
  checksum checksum
  mode node['mailhog']['binary']['mode']
  action :create
end

# Setup runit service
runit_service 'mailhog' do
  options({
    :smtp_ip => node['mailhog']['smtp']['ip'],
    :smtp_port => node['mailhog']['smtp']['port'],
    :smtp_outgoing => node['mailhog']['smtp']['outgoing'],
    :ui_ip => node['mailhog']['ui']['ip'],
    :ui_port => node['mailhog']['ui']['port'],
    :ui_web_path => node['mailhog']['ui']['web_path'],
    :api_ip => node['mailhog']['api']['ip'],
    :api_port => node['mailhog']['api']['port'],
    :cors_origin => node['mailhog']['cors-origin'],
    :hostname => node['mailhog']['hostname'],
    :storage => node['mailhog']['storage'],
    :maildir_path => node['mailhog']['maildir_path'],
    :mongodb_ip => node['mailhog']['mongodb']['ip'],
    :mongodb_port => node['mailhog']['mongodb']['port'],
    :mongodb_db => node['mailhog']['mongodb']['db'],
    :mongodb_collection => node['mailhog']['mongodb']['collection'],
    :jim_enable => node['mailhog']['jim']['enable'],
    :jim_accept => node['mailhog']['jim']['accept'],
    :jim_disconnect => node['mailhog']['jim']['disconnect'],
    :jim_linkspeed_affect => node['mailhog']['jim']['linkspeed']['affect'],
    :jim_linkspeed_max => node['mailhog']['jim']['linkspeed']['max'],
    :jim_linkspeed_min => node['mailhog']['jim']['linkspeed']['min'],
    :jim_reject_auth => node['mailhog']['jim']['reject']['auth'],
    :jim_reject_recipient => node['mailhog']['jim']['reject']['recipient'],
    :jim_reject_sender => node['mailhog']['jim']['reject']['sender']
  })
  default_logger true
  owner node['mailhog']['service']['owner']
  group node['mailhog']['service']['group']
  action [:enable, :restart]
end


attributes/default.rb
#
# Cookbook Name:: mailhog
# Recipe:: default
#
# Copyright (c) 2015 Sergey Storchay, All Rights Reserved.
# Modified 2016 Gleb Levitin, dkd Internet Service GmbH
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

default['mailhog']['version'] = '0.2.0'
default['mailhog']['install_method'] = 'binary'

default['mailhog']['binary']['url'] = nil # Set it to override automatical generation

default['mailhog']['binary']['mode'] = 0755
default['mailhog']['binary']['path'] = '/usr/local/bin/MailHog'

default['mailhog']['binary']['prefix_url'] = 'https://github.com/mailhog/MailHog/releases/download/v'
default['mailhog']['binary']['checksum']['linux_386'] = '06fc1d7bf2fced86962ab274d8f1e6c7df74a6ec3c6310aff26792bb93122d98'
default['mailhog']['binary']['checksum']['linux_amd64'] = '11aaae19388d0a6543c935188fcc4157772d852c10be2a5d96168ee23ae6845f'

default['mailhog']['service']['owner'] = 'root'
default['mailhog']['service']['group'] = 'root'

default['mailhog']['smtp']['ip'] = '127.0.0.1'
default['mailhog']['smtp']['port'] = 1025
default['mailhog']['smtp']['outgoing'] = nil

default['mailhog']['api']['ip'] = '0.0.0.0'
default['mailhog']['api']['port'] = 8025
default['mailhog']['ui']['ip'] = '0.0.0.0'
default['mailhog']['ui']['port'] = 8025
default['mailhog']['ui']['web_path'] = ''

default['mailhog']['cors-origin'] = nil
default['mailhog']['hostname'] = 'mailhog.example'

default['mailhog']['storage'] = 'memory'
default['mailhog']['maildir_path'] = ''
default['mailhog']['mongodb']['ip'] = '127.0.0.1'
default['mailhog']['mongodb']['port'] = 27017
default['mailhog']['mongodb']['db'] = 'mailhog'
default['mailhog']['mongodb']['collection'] = 'messages'

default['mailhog']['jim']['enable'] = false
default['mailhog']['jim']['accept'] = 0.99
default['mailhog']['jim']['disconnect'] = 0.005
default['mailhog']['jim']['linkspeed']['affect'] = 0.1
default['mailhog']['jim']['linkspeed']['max'] = 10240
default['mailhog']['jim']['linkspeed']['min'] = 1024
default['mailhog']['jim']['reject']['auth'] = 0.05
default['mailhog']['jim']['reject']['recipient'] = 0.05
default['mailhog']['jim']['reject']['sender'] = 0.05